### PR TITLE
IMP: expand support for metadata merging

### LINKED ISSUE #11: add `merge` action
**Comments**
Just like `merge_mapping_files.py`.

### PR DESCRIPTION
**Pull request by gregcaporaso**
This pull request is intended to be a starting point for more extensive metadata merging support in QIIME 2. This PR depends on https://github.com/qiime2/q2-types/pull/297.

Add support for merging metadata that contains overlapping ids, overlapping columns, neither overlapping ids nor overlapping columns, but not both overlapping ids an overlapping columns (because then there could be conflicts in column values for specific samples, which gets considerably more complex to handle). The result will be the union (i.e., outer join) of the ids and columns from the two metadata inputs.

This action currently only works on two metadata inputs at a time as providing an arbitrary number of metadata files at once would trigger the built-in metadata merging functionality (which would intersect the ids and fail on overlapping column names).

Example usage:
`md1.tsv`:
```
sample-id	barcode-sequence	body-site
#q2:types	categorical	categorical
L2S155	ACGATGCGACCA	left palm
L2S175	AGCTATCCACGA	left palm
L2S204	ATGCAGCTCAGT	left palm
L2S222	CACGTGACATGT	left palm
L3S242	ACAGTTGCGCGA	right palm
L3S294	CACGACAGGCTA	right palm
```

`md2.tsv`:
```
sample-id	body-site	year
#q2:types	categorical	numeric
L1S8	gut	2008
L1S57	gut	2009
L1S76	gut	2009
L1S105	gut	2009
```

```bash
$ qiime metadata merge --m-metadata1-file md1.tsv --m-metadata2-file md2.tsv --o-merged-metadata md.qza
$ qiime tools export --input-path md.qza --output-path md/
$ cat md/metadata.tsv
sample-id	barcode-sequence	body-site	year
#q2:types	categorical	categorical	numeric
L1S105		gut	2009
L1S57		gut	2009
L1S76		gut	2009
L1S8		gut	2008
L2S155	ACGATGCGACCA	left palm
L2S175	AGCTATCCACGA	left palm
L2S204	ATGCAGCTCAGT	left palm
L2S222	CACGTGACATGT	left palm
L3S242	ACAGTTGCGCGA	right palm
L3S294	CACGACAGGCTA	right palm
```

```bash
qiime metadata tabulate --m-input-file md.qza --o-visualization md.qzv
```

Screenshot 2023-07-13 at 7 07 40 AM

Fixes #11

### CODE REVIEW

File: q2_metadata/plugin_setup.py [hunk 1]
```diff
@@ -135,3 +136,31 @@ def _2(ff: ArtificialGroupingFormat) -> (qiime2.Metadata):
...
+    description=('Merge metadata that contains overlapping ids, overlapping '
+                 'columns, neither overlapping ids nor overlapping columns, '
+                 'but not both overlapping ids and overlapping columns. The '
+                 'result will be the union (i.e., outer join) of the '
+                 'ids and columns from the two metadata inputs.\n\nTo merge '
+                 'more than two metadata objects, run this command multiple '
+                 'times, iteratively using the output of the previous run as '
+                 'one of the metadata inputs.\n\n'
+                 'The output, an ImmutableMetadata artifact, can be used '
+                 'anywhere that a metafile file can be used, or can be '
```
Comment (lizgehret) [inline q2_metadata/plugin_setup.py:164]:
```suggestion
                 'anywhere that a metadata file can be used, or can be '
```

File: q2_metadata/plugin_setup.py [hunk 2]
```diff
@@ -135,3 +136,31 @@ def _2(ff: ArtificialGroupingFormat) -> (qiime2.Metadata):
...
+    parameter_descriptions={
+        'metadata1': 'First metadata file to merge.',
+        'metadata2': 'Second metadata file to merge.'
+    },
+    outputs=[('merged_metadata', ImmutableMetadata)],
+    output_descriptions={
+        'merged_metadata': 'The merged metadata.'
+    },
+    name='Merge metadata',
+    description=('Merge metadata that contains overlapping ids, overlapping '
```
Comment (lizgehret) [inline q2_metadata/plugin_setup.py:155]: I'm not exactly sure how to reword this sentence, but it sounded a bit weird to me initially and I had to re-read a couple of times (just the 'neither overlapping ids nor overlapping columns' bit).

**Review by lizgehret — COMMENTED** [review]
Couple of small inline comments, but overall this looks great! Going to pull down locally to test, but this looks like it should be basically ready to merge.

File: q2_metadata/plugin_setup.py (hunk 2 above)
Comment (gregcaporaso) [reply q2_metadata/plugin_setup.py]: Thanks for the input! I reworked this a bit to split the information into multiple sentences, and I think that should help. Let me know what you think.

Comment (lizgehret) [reply q2_metadata/plugin_setup.py]: That reads much better, thanks @gregcaporaso! I'll go ahead and get this one merged.

**Comment by gregcaporaso** [comment]
Thanks @lizgehret!

**Comment by lizgehret** [comment]
For posterity, this also resolves: https://github.com/qiime2/qiime2/issues/633

**Pull request merged**